backend/internal/service/user_service.go

```diff
@@ -630,6 +630,131 @@ func (s *UserService) disableUserInternal(ctx context.Context, userID string, tx
 		Error
 }
 
+func (s *UserService) CreateSignupToken(ctx context.Context, expiresAt time.Time, usageLimit int) (model.SignupToken, error) {
+	return s.createSignupTokenInternal(ctx, expiresAt, usageLimit, s.db)
+}
+
+func (s *UserService) createSignupTokenInternal(ctx context.Context, expiresAt time.Time, usageLimit int, tx *gorm.DB) (model.SignupToken, error) {
+	signupToken, err := NewSignupToken(expiresAt, usageLimit)
+	if err != nil {
+		return model.SignupToken{}, err
+	}
+
+	if err := tx.WithContext(ctx).Create(signupToken).Error; err != nil {
+		return model.SignupToken{}, err
+	}
+
+	return *signupToken, nil
+}
+
+func (s *UserService) SignupWithToken(ctx context.Context, token string, userData dto.UserCreateDto, ipAddress, userAgent string) (model.User, string, error) {
+	tx := s.db.Begin()
+	defer func() {
+		tx.Rollback()
+	}()
+
+	var signupToken model.SignupToken
+	err := tx.
+		WithContext(ctx).
+		Where("token = ?", token).
+		First(&signupToken).
+		Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return model.User{}, "", &common.TokenInvalidOrExpiredError{}
+		}
+		return model.User{}, "", err
+	}
+
+	if !signupToken.IsValid() {
+		return model.User{}, "", &common.TokenInvalidOrExpiredError{}
+	}
+
+	user, err := s.createUserInternal(ctx, userData, false, tx)
+	if err != nil {
+		return model.User{}, "", err
+	}
+
+	accessToken, err := s.jwtService.GenerateAccessToken(user)
+	if err != nil {
+		return model.User{}, "", err
+	}
+
+	s.auditLogService.Create(ctx, model.AuditLogEventAccountCreated, ipAddress, userAgent, user.ID, model.AuditLogData{
+		"signupToken": signupToken.Token,
+	}, tx)
+
+	signupToken.UsageCount++
+	err = tx.WithContext(ctx).Save(&signupToken).Error
+	if err != nil {
+		return model.User{}, "", err
+	}
+
+	// If usage limit reached, delete the token
+	if signupToken.IsUsageLimitReached() {
+		err = tx.WithContext(ctx).Delete(&signupToken).Error
+		if err != nil {
+			return model.User{}, "", err
+		}
+	}
```

Comment:
We can avoid an extra query if the token has reached its max capacity
```suggestion
	// If usage limit reached, delete the token
	if signupToken.IsUsageLimitReached() {
		err = tx.WithContext(ctx).Delete(&signupToken).Error
		if err != nil {
			return model.User{}, "", err
		}
	} else {
        err = tx.WithContext(ctx).Save(&signupToken).Error
        if err != nil {
            return model.User{}, "", err
        }
    }
```

Reply:
I have it after saving as if the usage count reaches its max after this most recent use it will delete it. Your logic wouldnt delete it till the next time or scheduled job. Which could lead to issues. Unless im missing something.

Reply:
No...

The current logic:

1. Always updates the count in the DB
2. After updating the DB, checks if the count has reached max (using the in-memory value) and then deletes it

So the update is performed always.

I'm suggesting skipping the update if we already know (because of the in-memory count) that the count has reached max

Reply:
Ahh yeah i see it now, i was confused. I can change it later.